execution_workflows/DaPars2/Dapars2_snakemake/workflow/scripts/results_to_bed.py

```diff
@@ -0,0 +1,182 @@
+                  cols_to_combine = ["transcript_id", "gene_symbol"],
+                  site_suffix = "prox",
+                  sep = "_",
+                  id_outcol = "name"):
+    '''
+    add an id column (named id_outcol),
+    ID created by combining values in columns specified in cols_to_combine & a suffix (site_suffix)
+    with values separated by '_'
+    '''
+
+    df["suffix"] = site_suffix
+
+    cols = cols_to_combine + ["suffix"]
+
+    df[id_outcol] = df[cols].apply(lambda row: sep.join(row.values.astype(str)), axis=1)
+
+    return df.drop(["suffix"], axis = 1)
+
+
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser(description = "Script to generate BED files of predicted proximal and distal poly(A) sites from DaPars2 output tables")
+
+    parser.add_argument("-i", "--input-tsvs", default='', help = "space separated string of file paths to TSV files generated by DaPars2_Multi_Sample<_Multi_Chr>.py", required=True)
+    parser.add_argument("-c","--coordinate-type", type=str, default="gff", choices=["bed","gff"], help="do regions specified in 'loci' column follow GFF or BED conventions? (default: %(default)s)")
+    parser.add_argument("-o", "--output-bed", type=str, default="dapars2_identified_polya_sites.bed", help="name of output BED file (default: %(default)s)")
+
+    args = parser.parse_args()
+
+    flist = list(args.input_tsvs.split(' '))
+    out_path = args.output_bed
+    # print(out_path)
+
+    if args.coordinate_type == "bed":
+        loci_one_based = False
+    elif args.coordinate_type == "gff":
+        loci_one_based = True
+    else:
+        raise ValueError("'-c'/'--coordinate-type' must be one of 'bed' or 'gff' - {} was provided".format(args.coordinate_type))
+
+
+    all_chrs = pd.concat([pd.read_csv(f, sep = "\t") for f in flist]).reset_index(drop=True)
+    # 1. Pull out transcript_id, gene_name, chromosome and strand from region ID
+    all_chrs[["transcript_id", "gene_symbol", "chromosome", "strand"]] = all_chrs["Gene"].str.split("|", expand = True)
+
+
+    # 2. Pull out predicted proximal poly(A) site and convert to BED coordinates
+    prox_all_chrs = proximal_site_to_bed_coords(all_chrs)
+
+    # 3. Pull out distal site from the 'last exon' used for each transcript
+    # e.g. loci column formatted like chr6:119923969-119926687
+    distal_all_chrs = distal_site_to_bed_coords(all_chrs,one_based=loci_one_based)
+
+    # 4. Assign an ID for each poly(A) site
+    prox_all_chrs = polya_site_id(prox_all_chrs, site_suffix = "proximal")
+    distal_all_chrs = polya_site_id(distal_all_chrs, site_suffix = "distal")
+
+    # 5. BED requires a score column (5th) - assign empty field to '.' (identification BED)
+    coords_all_chrs = pd.concat([prox_all_chrs, distal_all_chrs]).reset_index(drop=True)
+    coords_all_chrs["score"] = "."
+
+    # 6. Output to BED
+    bed_cols = ["chromosome", "start", "end", "name", "score", "strand"]
+
+    out_bed = coords_all_chrs[bed_cols]
```

Comment:
Would be good to add a drop_duplicates call (by start column), so duplicate predicted poly(A) sites don't appear in the output BED file. This theoretically could happen with DaPars which tests per transcript model e.g. if different transcripts share the same last exon